server/server/src/cold_chain/sensor.rs

```diff
@@ -0,0 +1,139 @@
+
+    match upsert_sensors(service_provider, store_id, &sensors).await {
+        Ok(response) => response,
+        Err(error) => HttpResponse::InternalServerError().body(format!("{:#?}", error)),
+    }
+}
+
+fn validate_input(sensors: &Vec<Sensor>) -> bool {
+    sensors.into_iter().all(|sensor| validate_sensor(sensor))
+}
+
+fn validate_sensor(sensor: &Sensor) -> bool {
+    let mac_regex = Regex::new(r"^([A-F0-9]{2}:){5}[A-F0-9]{2}( \| [\w]*)?$").unwrap();
+
+    if !mac_regex.is_match(&sensor.mac_address) {
+        return false;
+    }
+    if sensor.log_interval < 1 {
+        return false;
+    }
+    if sensor.programmed_date < 1 {
+        return false;
+    }
+    if sensor.name.len() < 1 {
+        return false;
+    }
+    if sensor.battery_level < 1 || sensor.battery_level > 100 {
+        return false;
+    }
+    match sensor.log_delay {
+        Some(log_delay) => {
+            if log_delay < 1 {
+                return false;
+            }
+        }
+        None => {}
+    };
+    true
+}
+
+async fn upsert_sensors(
+    service_provider: Data<ServiceProvider>,
+    store_id: Option<String>,
+    sensors: &Vec<Sensor>,
+) -> Result<HttpResponse, RepositoryError> {
+    let mut ctx = service_provider.basic_context()?;
+    if store_id.is_some() {
+        ctx.store_id = store_id.unwrap();
+    }
+    let service = &service_provider.sensor_service;
+
+    let mut results: Vec<repository::Sensor> = Vec::new();
+    sensors.into_iter().for_each(|sensor| {
+        if service.get_sensor(&ctx, sensor.id.clone()).is_ok() {
+            let sensor = UpdateSensor {
+                id: sensor.id.clone(),
+                name: Some(sensor.name.clone()),
+                is_active: Some(true),
+                location_id: None,
+                log_interval: Some(sensor.log_interval),
+                battery_level: Some(sensor.battery_level),
+            };
+            match service.update_sensor(&ctx, sensor) {
+                Ok(updated) => results.push(updated),
+                Err(e) => error!("Unable to insert sensor: {:#?}", e),
+            }
+        } else {
+            let sensor = InsertSensor {
+                id: sensor.id.clone(),
+                serial: sensor.mac_address.clone(),
+                name: Some(sensor.name.clone()),
+                is_active: Some(true),
+                log_interval: Some(sensor.log_interval),
+                battery_level: Some(sensor.battery_level),
+                r#type: get_sensor_type(&sensor.mac_address),
+            };
+            match service.insert_sensor(&ctx, sensor) {
+                Ok(inserted) => results.push(inserted),
+                Err(e) => error!("Unable to update sensor: {:#?}", e),
+            }
+        }
+    });
+
+    Ok(HttpResponse::Ok()
+        .append_header(header::ContentType(mime::APPLICATION_JSON))
+        .body(
+            serde_json::to_string(&results).unwrap_or("Unable to deserialise results".to_string()),
+        ))
+}
```

Comment:
.json(results) might be better.

Btw here and in the login endpoint, I think that response should only be mapped in top level abstraction, i.e. in sensors method (this way you can see from top level all of the operations and results)